torch/distributed/fsdp/_optim_utils.py

```diff
@@ -403,8 +404,17 @@ def _shard_orig_param_state(
             and value.dim() > 0
             and fsdp_state.sharding_strategy != ShardingStrategy.NO_SHARD
         ):
-            value = value.flatten()[intra_param_start_idx : intra_param_end_idx + 1].clone()  # type: ignore[operator]
+            if fsdp_state._device_mesh is not None:
+                # We have to call synchronize() if the tensor is gathered from
+                # DTensor.  Otherwise, the later `clone()` will cause errors.
```

Comment:
are you using the `full_tensor()` API to do the gathering or something else? iirc `full_tensor` gives sync behavior

Reply:
> are you using the `full_tensor()` API to do the gathering or something else? iirc `full_tensor` gives sync behavior

No. We are still using redistribute for the all_gather, since `full_tensor()` API was introduced later. 
Maybe we could update this line to use `full_tensor` API? https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_optim_utils.py#L1455